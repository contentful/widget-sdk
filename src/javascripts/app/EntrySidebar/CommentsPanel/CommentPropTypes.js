import PropTypes from 'prop-types';

export const Comment = PropTypes.shape({
  body: PropTypes.string.isRequired,
  sys: PropTypes.shape({
    id: PropTypes.string.isRequired,
    type: PropTypes.oneOf(['Comment']),
    commentType: PropTypes.oneOf(['Entry']),
    space: PropTypes.any,
    entry: PropTypes.object,
    createdBy: PropTypes.date,
    createdAt: PropTypes.string,
    updatedBy: PropTypes.object,
    updatedAt: PropTypes.string
  })
});

export const CommentThread = PropTypes.arrayOf(Comment);
