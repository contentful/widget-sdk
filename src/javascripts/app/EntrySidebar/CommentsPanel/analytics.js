import * as Intercom from 'services/intercom';

const INTERCOM_PREFIX = 'feature-comments';

export function trackCommentCreated() {
  Intercom.trackEvent(`${INTERCOM_PREFIX}-comment-created`);
}
