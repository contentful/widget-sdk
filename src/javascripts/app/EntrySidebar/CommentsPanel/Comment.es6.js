import React from 'react';
import moment from 'moment';
import { css } from 'emotion';
import {
  CardActions,
  DropdownList,
  DropdownListItem,
  Heading
} from '@contentful/forma-36-react-components';
import tokens from '@contentful/forma-36-tokens';

import * as types from './CommentPropTypes.es6';

export const styles = {
  comment: css({
    display: 'flex',
    flexDirection: 'column',
    marginBottom: tokens.spacingM,
    '&:last-child': {
      marginBottom: 0
    }
  }),
  header: css({
    display: 'flex',
    alignItems: 'flex-start',
    marginBottom: tokens.spacingM
  }),
  avatar: css({
    width: 36,
    height: 36,
    background: tokens.colorElementLight,
    borderRadius: '100%',
    marginRight: tokens.spacingS
  }),
  meta: css({
    display: 'flex',
    justifyContent: 'space-between',
    flexGrow: 1,
    flexDirection: 'column'
  }),
  userName: css({
    fontSize: tokens.fontSizeM
  }),
  timestamp: css({
    color: tokens.colorTextLightest,
    fontSize: tokens.fontSizeS
  }),
  commentBody: css({
    wordBreak: 'break-word',
    whiteSpace: 'pre-wrap'
  })
};

export default function Comment({ comment }) {
  const {
    sys: { createdBy, createdAt }
  } = comment;
  const creationDate = moment(createdAt, moment.ISO_8601);
  return (
    <div className={styles.comment}>
      <header className={styles.header}>
        <img className={styles.avatar} src={createdBy.avatarUrl} />
        <div className={styles.meta}>
          <Heading element="h4" className={styles.userName}>
            {`${createdBy.firstName} ${createdBy.lastName}`}
          </Heading>
          <time
            dateTime={creationDate.toISOString()}
            title={creationDate.format('LLLL')}
            className={styles.timestamp}>
            {creationDate.fromNow()}
          </time>
        </div>
        <CardActions>
          <DropdownList>
            <DropdownListItem onClick={() => {}}>Edit</DropdownListItem>
            <DropdownListItem onClick={() => {}}>Reply</DropdownListItem>
            <DropdownListItem onClick={() => {}}>Mark as resolved</DropdownListItem>
          </DropdownList>
          <DropdownList border="top">
            <DropdownListItem onClick={() => {}}>Remove</DropdownListItem>
          </DropdownList>
        </CardActions>
      </header>
      <div className={styles.commentBody}>{comment.body}</div>
    </div>
  );
}

Comment.propTypes = {
  comment: types.Comment.isRequired
};
