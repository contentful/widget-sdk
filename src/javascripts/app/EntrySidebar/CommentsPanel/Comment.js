import React, { useState } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { css, cx } from 'emotion';
import {
  Heading,
  Tooltip,
  Notification,
  ModalConfirm,
  Paragraph,
  IconButton
} from '@contentful/forma-36-react-components';
import tokens from '@contentful/forma-36-tokens';
import { remove as removeComment } from 'data/CMA/CommentsRepo';
import { canRemoveComment } from './utils';
import * as types from './CommentPropTypes';

export const styles = {
  comment: css({
    display: 'flex',
    flexDirection: 'column',
    marginBottom: tokens.spacingM,
    '&:last-child': {
      marginBottom: 0
    }
  }),
  header: css({
    display: 'flex',
    alignItems: 'flex-start',
    marginBottom: tokens.spacingM
  }),
  avatar: css({
    width: 36,
    height: 36,
    background: tokens.colorElementLight,
    borderRadius: '100%',
    marginRight: tokens.spacingS
  }),
  avatarSmall: css({
    width: 18,
    height: 18
  }),
  meta: css({
    display: 'flex',
    justifyContent: 'space-between',
    flexGrow: 1,
    flexDirection: 'column'
  }),
  userName: css({
    fontSize: tokens.fontSizeM
  }),
  timestamp: css({
    color: tokens.colorTextLight,
    fontSize: tokens.fontSizeS
  }),
  commentBody: css({
    wordBreak: 'break-word',
    whiteSpace: 'pre-wrap'
  }),
  commentBodyIndented: css({
    marginLeft: `calc(${tokens.spacingS} + 18px)` // For aligning comment body with avatar
  })
};

// TODO: Make this a dumb component like <Task /> neither triggering any CMA requests
//  directly, nor receiving an entire CMA `Comment` object.

export default function Comment({ endpoint, comment, onRemoved, className, hasReplies, isReply }) {
  const {
    sys: { createdBy, createdAt }
  } = comment;
  const creationDate = moment(createdAt, moment.ISO_8601);
  const handleRemove = async () => {
    const entry = comment.sys.reference || comment.sys.parentEntity;
    try {
      await removeComment(endpoint, entry.sys.id, comment);
      onRemoved(comment);
    } catch (err) {
      Notification.error(err.message);
    }
  };

  return (
    <div className={cx([styles.comment, className])} data-test-id="comment">
      <header className={styles.header}>
        <img
          className={cx(styles.avatar, isReply && styles.avatarSmall)}
          src={createdBy.avatarUrl}
          data-test-id="comment.avatar"
        />
        <div className={styles.meta}>
          <Heading element="h4" className={styles.userName} data-test-id="comment.user">
            {renderUserName(createdBy)}
          </Heading>
          <time
            dateTime={creationDate.toISOString()}
            title={creationDate.format('LLLL')}
            className={styles.timestamp}
            data-test-id="comment.timestamp">
            {creationDate.fromNow()}
          </time>
        </div>
        <CommentActions comment={comment} onRemove={handleRemove} hasReplies={hasReplies} />
      </header>
      <div
        className={cx(styles.commentBody, isReply && styles.commentBodyIndented)}
        data-test-id="comment.body">
        {comment.body}
      </div>
    </div>
  );
}

Comment.propTypes = {
  endpoint: PropTypes.func.isRequired,
  comment: types.Comment.isRequired,
  onRemoved: PropTypes.func.isRequired,
  className: PropTypes.string,
  // Only used to check if the comment can be removed.
  // Get rid of this once thread deletion is supported by the API
  hasReplies: PropTypes.bool,
  isReply: PropTypes.bool
};

function renderUserName(user) {
  return user.firstName ? (
    <React.Fragment>{`${user.firstName} ${user.lastName}`}</React.Fragment>
  ) : (
    <Tooltip content="The author of this comment is no longer a member of this organization">
      {'(Deactivated user)'}
    </Tooltip>
  );
}

function CommentActions({ comment, onRemove, hasReplies }) {
  const [showRemovalDialog, setShowRemovalDialog] = useState(false);

  if (hasReplies || !canRemoveComment(comment)) return null;

  return (
    <React.Fragment>
      <IconButton
        iconProps={{
          icon: 'Delete'
        }}
        buttonType="muted"
        testId="comment.menu.remove"
        label="Delete comment"
        onClick={() => setShowRemovalDialog(true)}
      />
      <RemovalConfirmationDialog
        isShown={showRemovalDialog}
        onConfirm={() => {
          setShowRemovalDialog(false);
          onRemove();
        }}
        onCancel={() => setShowRemovalDialog(false)}
        data-test-id="comment.removal-confirmation"
      />
    </React.Fragment>
  );
}
CommentActions.propTypes = {
  comment: types.Comment.isRequired,
  onRemove: PropTypes.func.isRequired,
  // Only used to check if the comment can be removed.
  // Get rid of this once thread deletion is supported by the API
  hasReplies: PropTypes.bool
};

function RemovalConfirmationDialog({ isShown, onConfirm, onCancel }) {
  return (
    <ModalConfirm
      isShown={isShown}
      title="Remove comment"
      intent="negative"
      onCancel={onCancel}
      onConfirm={onConfirm}>
      <Paragraph>Are you sure you want to remove this comment?</Paragraph>
    </ModalConfirm>
  );
}
RemovalConfirmationDialog.propTypes = {
  isShown: PropTypes.bool.isRequired,
  onConfirm: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired
};
