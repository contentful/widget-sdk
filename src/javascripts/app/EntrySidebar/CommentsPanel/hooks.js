import { useCallback } from 'react';
import { map, uniq } from 'lodash';
import resolveLinks from 'data/LinkResolver';
import { getSpace, getUserSync } from 'services/TokenStore';
import { createOrganizationEndpoint } from 'data/EndpointFactory';
import { getUsers } from 'access_control/OrganizationMembershipRepository';
import { getAllForEntry, create } from 'data/CMA/CommentsRepo';
import { useAsync, useAsyncFn } from 'core/hooks';
import { trackCommentCreated } from './analytics';

/**
 * Fetches all comments for the given entry.
 * Returns a state object
 * @param {SpaceEndpoint} endpoint
 * @param {string} entryId
 */
export const useCommentsFetcher = (endpoint, entryId) => {
  // avoiding infinite loop.
  const fetch = useCallback(async () => {
    return withMinDelay(fetchCommentsAndUsers(endpoint, entryId));
  }, [endpoint, entryId]);

  return useAsync(fetch);
};

/**
 * Creates a comment in the given entry.
 * Returns a state object and a requester function
 * @param {SpaceEndpoint} endpoint
 * @param {string} entryId
 */
export const useCommentCreator = (endpoint, entryId, parentCommentId) => {
  const user = getUserSync();

  return useAsyncFn(async (body) => {
    const comment = await create(endpoint, entryId, { body, parentCommentId });
    comment.sys.createdBy = user;
    trackCommentCreated();
    return comment;
  });
};

function withMinDelay(promise) {
  const delay = () => new Promise((resolve) => setTimeout(resolve, 700));
  return Promise.all([promise, delay()]).then(([data]) => data);
}

async function fetchUsers(spaceId, userIds) {
  const { organization } = await getSpace(spaceId);
  const orgEndpoint = createOrganizationEndpoint(organization.sys.id);
  // TODO: handle case where there are too many ids to fit into the query string
  const { items } = await getUsers(orgEndpoint, {
    'sys.id[in]': userIds.join(','),
  });
  return items;
}

/**
 * Get all comments as well as all the users associated with the comments
 * and return the resolved items
 */
export async function fetchCommentsAndUsers(endpoint, entryId) {
  const { spaceId } = endpoint;
  const { items: comments } = await getAllForEntry(endpoint, entryId);
  const commentCreatorIds = uniq(map(comments, 'sys.createdBy.sys.id'));
  const users = commentCreatorIds.length ? await fetchUsers(spaceId, commentCreatorIds) : [];
  const resolvedComments = resolveLinks({
    paths: ['sys.createdBy'],
    includes: { User: users },
    items: comments,
  });
  return resolvedComments;
}
