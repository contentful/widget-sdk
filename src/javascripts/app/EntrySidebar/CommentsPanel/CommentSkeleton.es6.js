import React from 'react';
import {
  Card,
  SkeletonContainer,
  SkeletonDisplayText,
  SkeletonImage,
  SkeletonBodyText
} from '@contentful/forma-36-react-components';
import { range } from 'lodash';
import { styles } from './Comment.es6';

export default function CommentSkeleton() {
  return (
    <Card className={styles.comment}>
      <SkeletonContainer svgHeight={100}>
        <SkeletonImage width={36} height={36} radiusX={36} radiusY={36} />
        <SkeletonDisplayText offsetTop={8} offsetLeft={50} width={80} />
        <SkeletonBodyText offsetTop={50} lineHeight={12} numberOfLines={2} />
      </SkeletonContainer>
    </Card>
  );
}

export function CommentSkeletonGroup({ number = 5 }) {
  return range(number).map(n => <CommentSkeleton key={n} />);
}
