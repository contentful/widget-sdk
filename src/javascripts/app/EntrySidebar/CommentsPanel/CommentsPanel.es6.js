import React from 'react';
import PropTypes from 'prop-types';
import SidebarEventTypes from '../SidebarEventTypes.es6';
import { css } from 'emotion';
import tokens from '@contentful/forma-36-tokens';
import { Card } from '@contentful/forma-36-react-components';
import Comment from './Comment.es6';
import CreateComment from './CreateEntryComment.es6';

const styles = {
  root: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    padding: '0',
    overflowY: 'auto',
    overflowX: 'hidden',
    color: tokens.colorTextMid,
    background: tokens.colorElementLightest,
    borderLeft: `1px solid ${tokens.colorElementDarkest}`,
    transition: 'transform .3s cubic-bezier(.38,.54,.5,.99)',
    display: 'flex',
    flexDirection: 'column',
    zIndex: 1
  },
  commentForm: {
    borderTop: `1px solid ${tokens.colorElementLight}`,
    padding: tokens.spacingS
  }
};

export default class CommentsPanel extends React.Component {
  static propTypes = {
    emitter: PropTypes.object.isRequired
  };

  state = {
    isVisible: true
  };

  componentDidMount() {
    this.props.emitter.on(SidebarEventTypes.UPDATED_COMMENTS_PANEL, this.updateVisibility);
  }

  componentWillUnmount() {
    this.props.emitter.off(SidebarEventTypes.UPDATED_COMMENTS_PANEL, this.updateVisibility);
  }

  updateVisibility = ({ isVisible }) => {
    this.setState({
      isVisible
    });
  };

  render() {
    const { isVisible } = this.state;

    return (
      <div
        className={css({
          ...styles.root,
          transform: isVisible ? 'translateX(-1px)' : 'translateX(100%)'
        })}>
        <div className={css({ overflow: 'auto', padding: tokens.spacingS })}>
          <Card className={css({ padding: tokens.spacingS, marginBottom: tokens.spacingS })}>
            <Comment />
          </Card>
          <Card className={css({ padding: tokens.spacingS, marginBottom: tokens.spacingS })}>
            <Comment />
          </Card>
        </div>
        <div className={css(styles.commentForm)}>
          <CreateComment />
        </div>
      </div>
    );
  }
}
