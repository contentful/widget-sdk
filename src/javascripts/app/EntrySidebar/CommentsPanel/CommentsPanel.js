import React, { useState, useEffect, useRef, useMemo, useCallback } from 'react';
import PropTypes from 'prop-types';
import { css } from 'emotion';

import tokens from '@contentful/forma-36-tokens';
import { Typography, Paragraph, Heading } from '@contentful/forma-36-react-components';

import { useCommentsFetcher } from './hooks';
import CreateComment from './CreateEntryComment';
import CommentThread from './CommentThread';
import CommentsPanelEmptyState from './CommentsPanelEmptyState';
import { CommentSkeletonGroup } from './CommentSkeleton';
import { fromFlatToThreads, isReply, isReplyToComment } from './utils';
import { useSpaceEnvEndpoint } from 'core/hooks/useSpaceEnvEndpoint';

export const styles = {
  root: css({
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    color: tokens.colorTextMid,
    background: tokens.colorElementLightest,
    transition: 'transform .3s cubic-bezier(.38,.54,.5,.99)',
    zIndex: 1,
  }),
  commentList: css({
    overflow: 'auto',
    flexGrow: 1,
    margin: `-${tokens.spacingM}`,
    padding: tokens.spacingM,
  }),
  commentListLoading: css({
    overflow: 'hidden',
  }),
  commentForm: css({
    borderTop: `1px solid ${tokens.colorElementLight}`,
    padding: `${tokens.spacingS} 0`,
    marginTop: tokens.spacingM,
  }),
  errorState: css({
    display: 'flex',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    textAlign: 'center',
  }),
};

function CommentsPanelContent({ entryId, onCommentsCountUpdate }) {
  const endpoint = useSpaceEnvEndpoint();

  const { isLoading, data, error } = useCommentsFetcher(endpoint, entryId);
  const [comments, setComments] = useState();
  const [shouldScroll, setShouldScroll] = useState(false);
  const listRef = useRef(null);

  const setUpdatedComments = useCallback(
    (comments) => {
      setComments(comments);
      onCommentsCountUpdate(comments.length);
    },
    [onCommentsCountUpdate]
  );

  const handleNewComment = (comment) => {
    const updatedComments = [...comments, comment];
    setUpdatedComments(updatedComments);
    !isReply(comment) && setShouldScroll(true);
  };

  // Remove a single comment or a comment and its replies
  const handleCommentRemoved = (comment) => {
    const isReply = isReplyToComment(comment);
    const newList = comments.filter((item) => {
      return !(item == comment || isReply(item));
    });
    setUpdatedComments(newList);
  };

  // check for the first load of comments.
  // display comments and scroll to the bottom
  // or display empty state
  useEffect(() => {
    if (data) {
      setUpdatedComments(data);
      setShouldScroll(true);
    }
  }, [data, setUpdatedComments]);

  useEffect(() => {
    // scroll to the bottom to show latest comments
    // after the first load or new comments.
    // the effect garantees that the scrolling happens after
    // the comments are rendered
    shouldScroll && scrollToBottom(listRef.current);
    setShouldScroll(false);
  }, [shouldScroll]);

  const isEmpty = !isLoading && !error && comments && comments.length === 0;
  const threads = useMemo(() => fromFlatToThreads(comments), [comments]);

  return (
    <React.Fragment>
      <div
        className={`${styles.commentList} ${isLoading ? styles.commentListLoading : ''}`}
        data-test-id="comments.list"
        ref={listRef}>
        {isLoading && <CommentSkeletonGroup />}
        {isEmpty && <CommentsPanelEmptyState />}
        {threads.length > 0 &&
          threads.map((thread) => (
            <CommentThread
              endpoint={endpoint}
              key={thread[0].sys.id}
              thread={thread}
              onNewReply={handleNewComment}
              onRemoved={handleCommentRemoved}
            />
          ))}
        {error && (
          <div className={styles.errorState} data-test-id="comments.error">
            <Typography>
              <Heading>Something went wrong</Heading>
              <Paragraph>{`We couldn't fetch the comments because of a problem.`}</Paragraph>
            </Typography>
          </div>
        )}
      </div>
      <div className={styles.commentForm} data-test-id="comments.form">
        <CreateComment endpoint={endpoint} entryId={entryId} onNewComment={handleNewComment} />
      </div>
    </React.Fragment>
  );
}

CommentsPanelContent.propTypes = {
  entryId: PropTypes.string.isRequired,
  onCommentsCountUpdate: PropTypes.func.isRequired,
};

export default function CommentsPanel({ isVisible, ...props }) {
  return (
    <div className={`${styles.root}`} data-test-id="comments">
      {isVisible && <CommentsPanelContent {...props} />}
    </div>
  );
}

CommentsPanel.propTypes = {
  ...CommentsPanelContent.propTypes,
  isVisible: PropTypes.bool,
};

function scrollToBottom(element) {
  try {
    // Chrome and Firefox
    element.scroll({
      top: element.scrollHeight,
      left: 0,
      behavior: 'smooth',
    });
  } catch {
    // Safari, IE and Edge
    element.scrollTop = element.scrollHeight;
  }
}
