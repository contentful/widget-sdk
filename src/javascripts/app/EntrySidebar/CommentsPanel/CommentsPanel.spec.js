import React from 'react';

import { render, within, fireEvent, screen, wait } from '@testing-library/react';

import CommentsPanel from './CommentsPanel';
import { useCommentsFetcher, useCommentCreator } from './hooks';
import { remove as removeComment } from 'data/CMA/CommentsRepo';

const mockAuthor = { firstName: 'John', lastName: 'Doe', avatarUrl: '0.jpeg', sys: { id: 'abc' } };

jest.mock('services/TokenStore', () => ({
  getSpace: jest.fn(),
  getUserSync: jest.fn().mockReturnValue({ sys: { id: 'abc' } })
}));
jest.mock('./hooks', () => ({
  useCommentsFetcher: jest.fn(),
  useCommentCreator: jest.fn().mockReturnValue([{}, jest.fn()])
}));
jest.mock('data/CMA/CommentsRepo', () => ({
  remove: jest.fn()
}));

const commentMeta = {
  createdBy: mockAuthor,
  space: { sys: { id: '123' } },
  reference: { sys: { id: 'xyz' } }
};

const mockComments = [
  { body: 'foobar1', sys: { id: '1', ...commentMeta } },
  { body: 'foobar2', sys: { id: '2', ...commentMeta } },
  { body: 'foobar3', sys: { id: '3', ...commentMeta } },
  { body: 'foobar4', sys: { id: '4', ...commentMeta } },
  { body: 'foobar5', sys: { id: '5', ...commentMeta } },
  { body: 'foobar6', sys: { id: '6', ...commentMeta } }
];

const mockReply = {
  body: 'foobar7',
  sys: { id: '7', parent: { sys: { id: '1' } }, ...commentMeta }
};

const setComments = (withReply = false) => {
  const comments = withReply ? [...mockComments, mockReply] : mockComments;
  useCommentsFetcher.mockReturnValue({ isLoading: false, error: null, data: comments });
};
const setError = () => {
  useCommentsFetcher.mockReturnValue({ isLoading: false, error: new Error('Oops'), data: null });
};
const setLoading = () => {
  useCommentsFetcher.mockReturnValue({ isLoading: true, error: null, data: null });
};
const setEmpty = () => {
  useCommentsFetcher.mockReturnValue({ isLoading: false, error: null, data: [] });
};

const defaultProps = {
  endpoint: () => {},
  entryId: 'b',
  environmentId: 'c',
  isVisible: true,
  onCommentsCountUpdate: jest.fn()
};

const build = (props = defaultProps) => {
  return render(<CommentsPanel {...props} />);
};

describe('CommentsPanel', () => {
  afterEach(useCommentsFetcher.mockClear);

  describe('with `isVisible: false`', () => {
    const props = {
      ...defaultProps,
      isVisible: false
    };

    it('does not fetch any comments', () => {
      build(props);
      expect(useCommentsFetcher).not.toHaveBeenCalled();
    });
  });

  describe('with `isVisible: true`', () => {
    beforeEach(setLoading);

    it('fetches all comments of entry `props.entryId` via `props.endpoint`', () => {
      build();
      const { endpoint, entryId } = defaultProps;
      expect(useCommentsFetcher).toHaveBeenCalledWith(endpoint, entryId);
    });
  });

  describe('loading', () => {
    beforeEach(setLoading);

    it('renders the loading state', () => {
      build();
      expect(screen.getByTestId('comments.loading')).toBeTruthy();
      expect(screen.queryByTestId('comments.error')).toBeNull();
    });
  });

  describe('empty', () => {
    beforeEach(setEmpty);

    it('renders the empty state', () => {
      build();
      expect(screen.getByTestId('comments.empty')).toBeTruthy();
      expect(screen.queryByTestId('comments.error')).toBeNull();
      expect(screen.queryByTestId('comments.loading')).toBeNull();
    });
  });

  describe('loaded', () => {
    it('renders the comments and calls comments count update callback', () => {
      setComments();
      build();
      expect(screen.queryAllByTestId('comments.thread')).toHaveLength(6);
      expect(defaultProps.onCommentsCountUpdate).toHaveBeenCalledWith(6);
    });

    it('renders threads and calls comments count update callback', () => {
      setComments(true);
      build();
      expect(screen.queryAllByTestId('comments.thread')).toHaveLength(6);
      expect(defaultProps.onCommentsCountUpdate).toHaveBeenCalledWith(7);
    });
  });

  describe('error', () => {
    beforeEach(setError);

    it('renders the error state', () => {
      build();
      expect(screen.getByTestId('comments.error')).toBeTruthy();
      expect(screen.queryByTestId('comments.empty')).toBeNull();
      expect(screen.queryByTestId('comments.loading')).toBeNull();
    });
  });

  // From this point on, tests require knowledge about the child components
  // You can make the case that they are all part of the same thing.
  // For instance, the only way for a user to create a comment is by using the form
  describe('comment added', () => {
    const addComment = container => {
      const comment = {
        body: 'foobar8',
        sys: { id: '8', createdBy: mockAuthor, parentEntity: { sys: { id: 'test' } } }
      };
      const form = within(container).getByTestId('comments.form');
      const textarea = within(form).getByTestId('comments.form.textarea');
      fireEvent.focus(textarea); // focus to show buttons
      const submitBtn = within(form).getByTestId('comments.form.submit');
      fireEvent.change(textarea, { target: { value: 'My new comment' } });
      useCommentCreator.mockReturnValueOnce([
        { isLoading: false, error: null, data: comment },
        jest.fn()
      ]);
      fireEvent.click(submitBtn);
      return wait();
    };

    it('adds a new comment when the list is empty and calls comments count update callback', () => {
      setEmpty();
      const { container } = build();
      expect(screen.queryAllByTestId('comments.thread')).toHaveLength(0);
      addComment(container);
      expect(screen.queryByTestId('comments.empty')).toBeNull();
      expect(screen.queryAllByTestId('comments.thread')).toHaveLength(1);
      expect(defaultProps.onCommentsCountUpdate).toHaveBeenCalledWith(1);
    });

    it('adds a new comment when the list has other comments and calls comments count update callback', () => {
      setComments();
      const { container } = build();
      expect(screen.queryAllByTestId('comments.thread')).toHaveLength(6);
      addComment(container);
      expect(screen.queryAllByTestId('comments.thread')).toHaveLength(7);
      expect(defaultProps.onCommentsCountUpdate).toHaveBeenCalledWith(7);
    });
  });

  describe('comment deleted', () => {
    const deleteComment = container => {
      const removeButton = within(container).getAllByTestId('comment.menu.remove')[0];
      fireEvent.click(removeButton);

      removeComment.mockResolvedValue();

      const confirmButton = screen.getByTestId('cf-ui-modal-confirm-confirm-button');
      fireEvent.click(confirmButton);
      return wait();
    };

    it('deletes a comment when the list has other comments and calls comments count update callback', async () => {
      setComments();
      const { container } = build();
      expect(screen.getAllByTestId('comments.thread')).toHaveLength(6);
      await deleteComment(container);
      expect(screen.getAllByTestId('comments.thread')).toHaveLength(5);
      expect(defaultProps.onCommentsCountUpdate).toHaveBeenCalledWith(5);
    });
  });
});
