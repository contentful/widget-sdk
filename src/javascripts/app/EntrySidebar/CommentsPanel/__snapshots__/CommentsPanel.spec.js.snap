// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsPanel error matches the snapshot 1`] = `
.emotion-2 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  padding: 0;
  overflow-y: auto;
  overflow-x: hidden;
  color: #536171;
  background: #f7f9fa;
  border-left: 1px solid #b4c3ca;
  -webkit-transition: -webkit-transform .3s cubic-bezier(.38,.54,.5,.99);
  -webkit-transition: transform .3s cubic-bezier(.38,.54,.5,.99);
  transition: transform .3s cubic-bezier(.38,.54,.5,.99);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  z-index: 1;
}

.emotion-3 {
  -webkit-transform: translateX(100%);
  -ms-transform: translateX(100%);
  transform: translateX(100%);
}

.emotion-0 {
  overflow: auto;
  padding: 0.75rem;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.emotion-1 {
  border-top: 1px solid #e5ebed;
  padding: 0.75rem;
}

<div
  className="emotion-2 emotion-3"
>
  <div
    className="emotion-0"
  />
  <div
    className="emotion-1"
  >
    <CreateEntryComment
      entryId="b"
      onNewComment={[Function]}
      spaceId="a"
    />
  </div>
</div>
`;

exports[`CommentsPanel loading matches the snapshot 1`] = `
.emotion-2 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  padding: 0;
  overflow-y: auto;
  overflow-x: hidden;
  color: #536171;
  background: #f7f9fa;
  border-left: 1px solid #b4c3ca;
  -webkit-transition: -webkit-transform .3s cubic-bezier(.38,.54,.5,.99);
  -webkit-transition: transform .3s cubic-bezier(.38,.54,.5,.99);
  transition: transform .3s cubic-bezier(.38,.54,.5,.99);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  z-index: 1;
}

.emotion-3 {
  -webkit-transform: translateX(100%);
  -ms-transform: translateX(100%);
  transform: translateX(100%);
}

.emotion-0 {
  overflow: auto;
  padding: 0.75rem;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.emotion-1 {
  border-top: 1px solid #e5ebed;
  padding: 0.75rem;
}

<div
  className="emotion-2 emotion-3"
>
  <div
    className="emotion-0"
  >
    <CommentSkeletonGroup />
  </div>
  <div
    className="emotion-1"
  >
    <CreateEntryComment
      entryId="b"
      onNewComment={[Function]}
      spaceId="a"
    />
  </div>
</div>
`;
